### ISSUE TITLE
cvars.AddChangeCallback doesn't work for all convars

### ISSUE THREAD

**Issue by Kefta** [Facepunch/garrysmod-issues#3503]
Try skill or gmod_language serverside.

**Comment by robotboy655** [comment]
But It does work for me?

![image](https://user-images.githubusercontent.com/3299036/41598373-abd2f1cc-73d8-11e8-9fc0-03f0436004eb.png)

**Comment by Kefta** [comment]
I specifically said serverside.

**Comment by robotboy655** [comment]
How do you expect clientside console variables to be detected serverside?

**Comment by Kefta** [comment]
skill exists serverside: https://github.com/VSES/SourceEngine2007/blob/43a5c90a5ada1e69ca044595383be67f40b33c61/src_main/engine/host.cpp#L478
Same with gmod_language

**Comment by Kefta** [comment]
They exist shared but are not replicated.

**Comment by meepen** [comment]
This doesn't work at all in menu state, same with fcvar archive

**Comment by Kefta** [comment]
@meepen https://github.com/Facepunch/garrysmod-issues/issues/1440

**Comment by meepen** [comment]
yes i'm saying it's probably the same issue

**Comment by Kefta** [comment]
Can this be fixed since non-menu states are officially moddable?

**Comment by robotboy655** [comment]
Looked into it, and its this way on purpose.

ConVars **without** ```FCVAR_GAMEDLL``` will not work on server and **with** ```FCVAR_GAMEDLL``` will not work on client. This does not affect Lua ConVars.

Not sure why it is this way, and I am not sure we want to change it.

**Comment by Kefta** [comment]
I don't see how altering this behaviour will affect current code and thus I think it should be changed. What purpose does this condition serve other than restricting valid convars from being monitored? It just leads to [hacky code](https://github.com/Kefta/gs_skilllevel/blob/master/lua/autorun/server/gs_skilllevel.lua#L175-L183) to achieve the same purpose as ``cvars.AddChangeCallback.``

**Comment by neico** [comment]
This restriction isn't present in native convar code (unless I've overlooked it),  
`FCVAR_GAMEDLL` marks something server only
`FCVAR_CLIENTDLL` marks something client only
not specifying either means that it can be used on both sides (shared, not to be confused with `FCVAR_REPLICATED` which still only allows one callback as replicated makes the cvar server side with a forced shadow copy for each client), but it can never be used to traverse from one side to another.

There can only ever be one callback for each convar (and a global callback) at any given time, so if it's to prevent to override a existing one then it should be handled accordingly, not like this.  
If it's to prevent interacting with native convars you can check for `FCVAR_LUA_SERVER` and `FCVAR_LUA_CLIENT` (why's there no `FCVAR_LUA_MENU`, or if there is, why is it not documented on the wiki?) to restrict it properly, again, not like this.

I don't see why we need to artificially limit of what is possible here,  
if overriding is a issue you can always modify the system to work more like the existing hook system and give native callbacks priority over lua created ones?

**Comment by Kefta** [comment]
Any update on this?

**Comment by robotboy655** [comment]
My main concern would be this being yet another thing that fires `C > Lua` every frame potentially affecting performance.

There are certain engine convars that are set every frame, like the HDR stuff.

**Comment by Kefta** [comment]
Maybe it can be formatted like ENTITY.RenderOverride or AddCallback so that the C->Lua call is optional.

**Comment by LeQuackers** [comment]
Yeah I'm with Kefta. You could mark the convars with a custom flag (maybe ```FCVAR_LUA_NOTIFY```?) and check if the flag is set in a similar manner to [```FCVAR_NOTIFY```](https://github.com/VSES/SourceEngine2007/blob/master/src_main/engine/cvar.cpp#L72).

**Comment by robotboy655** [comment]
This was a GMod specific behavior to try to funnel correct convars into correct realms. 

I have made it call the callback for both client and server if the convar has no sv/cl identifying flags.

This does mean that you will get a whole bunch of clientside only commands on the server realm now, that's the trade off. I could maybe use `cl_` and `mat_`, etc prefixes in addition to the flags, but it feels like an error prone hack

I am not going through the entire engine and updating 3000+ convars with correct flags.

**Comment by meepen** [comment]
(perfect use case for ai)
